RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitFromSoAAlpaka.cc

```diff
@@ -0,0 +1,185 @@
+      hostPutToken_(produces()) {}
+
+template <typename TrackerTraits>
+void SiPixelRecHitFromSoAAlpaka<TrackerTraits>::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("pixelRecHitSrc", edm::InputTag("siPixelRecHitsPreSplittingAlpaka"));
+  desc.add<edm::InputTag>("src", edm::InputTag("siPixelClustersPreSplitting"));
+  descriptions.addWithDefaultLabel(desc);
+}
+
+template <typename TrackerTraits>
+void SiPixelRecHitFromSoAAlpaka<TrackerTraits>::produce(edm::StreamID streamID,
+                                                        edm::Event& iEvent,
+                                                        const edm::EventSetup& iSetup) const {
+  auto const& hits = iEvent.get(hitsToken_);
+  auto nHits = hits.view().metadata().size();
+  LogDebug("SiPixelRecHitFromSoAAlpaka") << "converting " << nHits << " Hits";
+
+  // allocate a buffer for the indices of the clusters
+  constexpr auto nMaxModules = TrackerTraits::numberOfModules;
+
+  SiPixelRecHitCollection output;
+  output.reserve(nMaxModules, nHits);
+
+  HMSstorage hmsp(nMaxModules + 1);
+
+  if (0 == nHits) {
+    hmsp.clear();
+    iEvent.emplace(rechitsPutToken_, std::move(output));
+    iEvent.emplace(hostPutToken_, std::move(hmsp));
+    return;
+  }
+
+  // fill content of HMSstorage product, and put it into the Event
+  for (unsigned int idx = 0; idx < hmsp.size(); ++idx) {
+    hmsp[idx] = hits.view().hitsModuleStart()[idx];
+  }
+  iEvent.emplace(hostPutToken_, std::move(hmsp));
+
+  auto xl = hits.view().xLocal();
+  auto yl = hits.view().yLocal();
+  auto xe = hits.view().xerrLocal();
+  auto ye = hits.view().yerrLocal();
+
+  TrackerGeometry const& geom = iSetup.getData(geomToken_);
+
+  auto const hclusters = iEvent.getHandle(clusterToken_);
+
+  constexpr uint32_t maxHitsInModule = pixelClustering::maxHitsInModule();
+
+  int numberOfDetUnits = 0;
+  int numberOfClusters = 0;
+  for (auto const& dsv : *hclusters) {
+    numberOfDetUnits++;
+    unsigned int detid = dsv.detId();
+    DetId detIdObject(detid);
+    const GeomDetUnit* genericDet = geom.idToDetUnit(detIdObject);
+    auto gind = genericDet->index();
+    const PixelGeomDetUnit* pixDet = dynamic_cast<const PixelGeomDetUnit*>(genericDet);
+    assert(pixDet);
+    SiPixelRecHitCollection::FastFiller recHitsOnDetUnit(output, detid);
+    auto fc = hits.view().hitsModuleStart()[gind];
+    auto lc = hits.view().hitsModuleStart()[gind + 1];
+    auto nhits = lc - fc;
+
+    assert(lc > fc);
+    LogDebug("SiPixelRecHitFromSoAAlpaka") << "in det " << gind << ": conv " << nhits << " hits from " << dsv.size()
+                                           << " legacy clusters" << ' ' << fc << ',' << lc << "\n";
+    if (nhits > maxHitsInModule)
+      edm::LogWarning("SiPixelRecHitFromSoAAlpaka") << fmt::sprintf(
+          "Too many clusters %d in module %d. Only the first %d hits will be converted", nhits, gind, maxHitsInModule);
```

Comment:
How about
```suggestion
      edm::LogWarning("SiPixelRecHitFromSoAAlpaka").format(
          "Too many clusters {} in module {}. Only the first {} hits will be converted", nhits, gind, maxHitsInModule);
```
? Then the `fmt/printf.h` include. could be removed.